www/source/blog/2017-10-24-node-with-db.html.md

```diff
@@ -0,0 +1,348 @@
+
+pkg_binds=(
+  [database]="port"
+)
+```
+
+Now, let's define the value for that port in a configuration file - this will tell Habitat where to connect to the database when binding the two containers (unlike habitat/default.toml - which is used to set environmental variables within the application itself).
+
+**default-config.json**
+
+```
+{
+  "mongo": {
+    "host" : "localhost",
+    "port"   : "27017"
+  }
+}
+```
+
+Let's add one more thing to the plan.sh file - when we run the application in a Docker container, we need a port to access it on.
+
+**habitat/plan.sh**
+
+```
+pkg_name=learn-about-me
+pkg_origin=core
+pkg_version="0.1.0"
+pkg_scaffolding="core/scaffolding-node"
+
+pkg_binds=(
+  [database]="port"
+)
+
+pkg_exports=(
+  [port]=app.port
+)
+```
+
+We will define this port when we start up the container running our application - we'll get to that in just a moment.
+
+## Habitizing your application - Part II
+
+We need to make a few edits to the application code itself in order to make this work with Habitat. Open up your app.js file, the beginning of it should look like this:
+
+**app.js**
+
+```
+var bodyParser = require("body-parser");
+var cookieParser = require("cookie-parser");
+var express = require("express");
+var flash = require("connect-flash");
+var mongoose = require("mongoose");
+var passport = require("passport");
+var path = require("path");
+var session = require("express-session");
+
+var setUpPassport = require("./setuppassport");
+var routes = require("./routes");
+
+var app = express();
+mongoose.connect("mongodb://localhost:27017/test");
+setUpPassport();
+```
+
+First, delete this line - since we define the database url through Habitat:
+
+```
+mongoose.connect("mongodb://localhost:27017/test");
+```
+
+Now, add some content right after this line:
+
+```
+var app = express();
+```
+
+Like so:
+
+**app.js**
+
+```
+var app = express();
+var nconf = require('nconf');
+const nconf_file = process.env.APP_CONFIG || './default-config.json';
+nconf.file({ file: nconf_file });
+
+console.log("Loading Configuration");
+
+console.log(nconf.get('mongo'));
+var dbConnectionString = 'mongodb://' + nconf.get('mongo')['host'] + ':' + nconf.get('mongo')['port'] + "/test'";
+console.log(dbConnectionString);
+
+console.log("Connecting to Database");
+
+var mongoose = require('mongoose');
+mongoose.connect(dbConnectionString, { useMongoClient: true });
+```
+
+Wow, that's a lot.  Let's break this down:
+
+```
+var nconf = require('nconf');
+```
+
+First, we use an npm module called [nconf](https://www.npmjs.com/package/nconf).  Nconf is a tool to define configuration variables, then use them within the node application.
+
+
+```
+const nconf_file = process.env.APP_CONFIG;
+nconf.file({ file: nconf_file });
+```
+
+Then we read a configuration file and load it to nconf. This brings us to the question - where do we get this file?
+
+Let's create a template for this config file at habitat/config/config.json
```

Comment:
Considering the above code:
```
const nconf_file = process.env.APP_CONFIG || './default-config.json';
```

Because it doesn't mention this file at all and this file is not mentioned until you have installed a dependency and updated the plan file. As a learner I would have a hard time making the connection.